### PR TITLE
GUI support for Contact Settings

### PR DESCRIPTION
**Pull request by cjb**
@keybase/picnicsquad CC @keybase/react-hackers @keybase/design 

Here's a first shot at the contact settings screen.  Some details:

Screen Shot 2019-12-18 at 10 28 35 AM

Screen Shot 2019-12-18 at 11 34 32 AM

* The first checkbox ("Only let someone..") can be toggled and saved.

* The second checkbox ("I follow them..") can't be toggled -- it's always on, just there to inform about what's going to be happening.  (I thought about making the checkbox fully disabled, but that makes the text have lower opacity too and looks weird.)

* The third checkbox ("I follow someone..") can be toggled and saved.

* The fourth checkbox ("They're in one..") can be toggled, but the GUI doesn't remember whether you set it, so it'll always be unset when you return to the tab.  Can be fixed on the server side but hasn't been yet.

* The choices in the team list can currently be saved but not reloaded, so it'll look like your choices weren't persisted even though they were.  Max has a PR up to fix that.

### CODE REVIEW

File: shared/settings/chat/index.tsx [hunk 1]
```diff
@@ -70,9 +93,102 @@ class Chat extends React.Component<Props, State> {
...
+      this.setState({contactSettingsIndirectFollowees: this.props.contactSettingsIndirectFollowees})
+    }
+    // Create an initial copy of teams data into state, so it can be mutated there.
+    if (
+      Object.keys(this.props.contactSettingsSelectedTeams).length > 0 &&
+      Object.keys(this.state.contactSettingsSelectedTeams).length == 0
+    ) {
+      this.setState({contactSettingsSelectedTeams: this.props.contactSettingsSelectedTeams})
+    }
+  }
```
Comment (thebearjew) [inline shared/settings/chat/index.tsx]: This comment is more of an exploration of whether I understand hooks correctly, and less a serious suggested change.

Could this be a functional component with hooks? `useState` and `useEffect` should work here?


```js
render() {
    const [contactSettingsEnabled, setContactSettingsEnabled] = React.useState(undefined)
    const [contactSettingsIndirectFollowees, setSontactSettingsIndirectFollowees] = React.useState(undefined)
    const [contactSettingsTeamsEnabled, setSontactSettingsTeamsEnabled] = React.useState(undefined)
    const [contactSettingsSelectedTeams, setContactSettingsSelectedTeams] = React.useState({})
    React.useEffect(() => {
        setContactSettingsSelectedTeams(props.contactSettingsSelectedTeams)
    }, [Object.keys(props.contactSettingsSelectedTeams).length > 0
        && Object.keys(contactSettingsSelectedTeams).length == 0])
    return (
        ...
        // Later - can use the `prevState` function
        <TeamRow
            ...
            onCheck={(checked: boolean) => setContactSettingsSelectedTeams(
                prevState => ({ ...prevState, [teamDetails.id]: checked })
            }
        />
    )
}
```

File: shared/settings/chat/index.tsx [hunk 2]
```diff
@@ -70,9 +93,102 @@ class Chat extends React.Component<Props, State> {
...
+                  !!this.state.contactSettingsEnabled,
+                  !!this.state.contactSettingsIndirectFollowees,
+                  !!this.state.contactSettingsTeamsEnabled,
+                  this.state.contactSettingsSelectedTeams
+                )
+              }
+              label="Save"
+              style={styles.save}
+              waitingKey={Constants.contactSettingsWaitingKey}
+            />
```
Comment (thebearjew) [inline shared/settings/chat/index.tsx]: I think this save button is a bit smushed between the post link previews and the teams list above. 
Would it be possible to add some more space above/below this button?

File: shared/settings/chat/index.tsx [hunk 3]
```diff
@@ -1,24 +1,47 @@
...
+  onUnfurlSave: (mode: RPCChatTypes.UnfurlMode, whitelist: Array<string>) => void
   onRefresh: () => void
+  teamDetails: Array<TeamDetails>
 }
 
 type State = {
+  contactSettingsEnabled?: boolean | null
+  contactSettingsIndirectFollowees?: boolean | null
+  contactSettingsSelectedTeams: {[K in TeamID]: boolean}
+  contactSettingsTeamsEnabled?: boolean | null
```
Comment (thebearjew) [inline shared/settings/chat/index.tsx]: Can't seem to understand when/why/where these would be null? state s initialized to `undefined` and I think the RPCs don't return `null`. Could be wrong here.

Comment (cjb) [reply shared/settings/chat/index.tsx]: @thebearjew Ah, I think it's the case where the load RPC failed and we set `contactSettingsError` but nothing else.

File: shared/settings/chat/index.tsx (hunk 1 above)
Comment (cjb) [reply shared/settings/chat/index.tsx]: Looks right!  I thought I'd keep it in line with the way Unfurl works too.

**Comment by cecileboucheron** [comment]
Some design feedback:
1. Could we nest the 3 checkboxes under the very top level one on the x axis, so it's clear they're children?
2. Let's space out the switches on mobile vertically - they seem too close to each other (for fat fingers)
3. the screenshots don't show: does the team list show up only if the team filtering checkbox is checked? do the 3 children boxes show up only if the top parent box is checked?
4. Could we do smaller team avatars, and hide the number of members - so it better fits within the vertical space of the box/switch
5. remove container and nested scroll - let's avoid vertical scrolls inside vertical scrolls, especially on mobile
6. the spacing to the right of the team switches should be the same than the spacing to the left

**Comment by malgorithms** [comment]
> 1. Could we nest the 3 checkboxes under the very top level one on the x axis, so it's clear they're children?

agree strongly with this one - those have to feel like an indented subsetting of that checkbox

**Comment by cecileboucheron** [comment]
Also let's switch to `you` everywhere - we generally don't use I/me in the app: 
`Only let someone message you or add you to a team if:
You follow them, or...
You follow someone who follows them, or
They're in one of these teams with you:`

**Comment by cjb** [comment]
@cecileboucheron Hi!  How's this?  I've made the changes you asked for above, and it is the case that the subsequent checkboxes only appear after the first one's checked, and the team list only appears after that checkbox is checked.

Screen Shot 2019-12-31 at 11 50 41 AM

Screen Shot 2019-12-31 at 12 19 35 PM

**Comment by cecileboucheron** [comment]
Looking good!

**Comment by cjb** [comment]
@thebearjew Thanks, good call, added a story for each of the option combinations.

**Pull request merged**